processing/src/main/java/io/druid/query/CacheStrategy.java

```diff
@@ -61,14 +61,16 @@
    *
    * The resulting function must be thread-safe.
    *
+   * @param isResultLevelCache indicates whether the function is invoked for result-level caching or segment-level caching
    * @return a thread-safe function that converts the QueryType's result type into something cacheable
    */
-  Function<T, CacheType> prepareForCache();
+  Function<T, CacheType> prepareForCache(boolean isResultLevelCache);
 
   /**
    * A function that does the inverse of the operation that the function prepareForCache returns
    *
+   * @param isResultLevelCache indicates whether the function is invoked for result-level caching or segment-level caching
    * @return A function that does the inverse of the operation that the function prepareForCache returns
    */
-  Function<CacheType, T> pullFromCache();
+  Function<CacheType, T> pullFromCache(boolean isResultLevelCache);
```

Comment:
Since a lot of calls default to `false`, would it make sense to add another method that just calls `pullFromCache(false)` and `prepareForCache(false)` ? and preserve backwards compat?